compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -70,6 +72,77 @@ EncodingAttr getEncodingAttr(RankedTensorType type) {
   return type.getEncoding().dyn_cast_or_null<EncodingAttr>();
 }
 
+AffineMap getMapForRole(EncodingAttr encoding) {
+  EncodingRole role = encoding.getRole().getValue();
+  if (role == EncodingRole::LHS)
+    return cast<AffineMapAttr>(encoding.getIndexingMaps()[0]).getAffineMap();
+  else if (role == EncodingRole::RHS)
+    return cast<AffineMapAttr>(encoding.getIndexingMaps()[1]).getAffineMap();
+  else
+    return cast<AffineMapAttr>(encoding.getIndexingMaps()[2]).getAffineMap();
+}
+
+FailureOr<linalg::ContractionDimensions>
+getEncodingContractionDims(EncodingAttr encoding) {
+  auto indexingMapsAttr = encoding.getIndexingMaps();
+  SmallVector<AffineMap> indexingMaps = llvm::map_to_vector(
+      indexingMapsAttr.getValue(), [](Attribute m) -> AffineMap {
+        return cast<AffineMapAttr>(m).getAffineMap();
+      });
+  return linalg::inferContractionDims(indexingMaps);
+}
+
+unsigned getResultIndex(AffineMap m, int64_t idx) {
+  return m.getResultPosition(getAffineDimExpr(idx, m.getContext())).value();
+}
+
+/// Get the permutation that permutes the input shape to the canonical
+/// matmul input shape based on the IndexingMaps encoding attribute.
+std::optional<SmallVector<int64_t>>
+getPermutationToCanonicalMatmulShape(EncodingAttr encoding) {
+  FailureOr<linalg::ContractionDimensions> cDims =
+      getEncodingContractionDims(encoding);
+  if (failed(cDims)) {
+    return std::nullopt;
+  }
+  SmallVector<int64_t> perm;
+  EncodingRole role = encoding.getRole().getValue();
+  EncodingUser user = encoding.getUser().getValue();
+  AffineMap map = getMapForRole(encoding);
+  // Add batch dim
+  if (user == EncodingUser::BATCH_MATMUL) {
+    perm.push_back(getResultIndex(map, cDims->batch[0]));
+  }
+  // Add M dim
+  if (role != EncodingRole::RHS && !isVecmatEncoding(encoding) &&
+      !isBatchVecmatEncoding(encoding)) {
+    perm.push_back(getResultIndex(map, cDims->m[0]));
+  }
+  // Add K dim
+  if (role != EncodingRole::RESULT) {
+    perm.push_back(getResultIndex(map, cDims->k[0]));
+  }
+  // Add N dim
+  if (role != EncodingRole::LHS && !isMatvecEncoding(encoding) &&
+      !isBatchMatvecEncoding(encoding)) {
+    perm.push_back(getResultIndex(map, cDims->n[0]));
+  }
```

Comment:
This comment can be a follow-up.

One of the goals is to remove `EncodingUser` from encodings. Are we able to just check if `cDims->batch[0]` is not empty?

Also, do we still need the check for vecmat and matvec? Ideally, we can get the information from indexing_maps?